tsdb/engine/tsm1/encoding.go

```diff
@@ -0,0 +1,571 @@
+// makes the code cleaner.
+type Values []Value
+
+func (v Values) MinTime() int64 {
+	return v[0].Time().UnixNano()
+}
+
+func (v Values) MaxTime() int64 {
+	return v[len(v)-1].Time().UnixNano()
+}
+
+func (v Values) Encode(buf []byte) ([]byte, error) {
+	switch v[0].(type) {
+	case *FloatValue:
+		a := make([]*FloatValue, len(v))
+		for i, vv := range v {
+			a[i] = vv.(*FloatValue)
+		}
+		return encodeFloatBlock(buf, a)
+
+	case *Int64Value:
+		a := make([]*Int64Value, len(v))
+		for i, vv := range v {
+			a[i] = vv.(*Int64Value)
+		}
+		return encodeInt64Block(buf, a)
+
+	case *BoolValue:
+		a := make([]*BoolValue, len(v))
+		for i, vv := range v {
+			a[i] = vv.(*BoolValue)
+		}
+		return encodeBoolBlock(buf, a)
+
+	case *StringValue:
+		a := make([]*StringValue, len(v))
+		for i, vv := range v {
+			a[i] = vv.(*StringValue)
+		}
+		return encodeStringBlock(buf, a)
+	}
+
+	return nil, fmt.Errorf("unsupported value type %T", v[0])
+}
+
+func (v Values) DecodeSameTypeBlock(block []byte) Values {
+	switch v[0].(type) {
+	case *FloatValue:
+		a, _ := decodeFloatBlock(block)
+		return a
+	case *Int64Value:
+		a, _ := decodeInt64Block(block)
+		return a
+	case *BoolValue:
+		a, _ := decodeBoolBlock(block)
+		return a
+	case *StringValue:
+		a, _ := decodeStringBlock(block)
+		return a
+	}
+	return nil
+}
+
+// DecodeBlock takes a byte array and will decode into values of the appropriate type
+// based on the block
+func DecodeBlock(block []byte) (Values, error) {
+	if len(block) == 0 {
+		return Values{}, nil
+	}
+
+	blockType := block[8]
+	switch blockType {
+	case BlockFloat64:
+		return decodeFloatBlock(block)
+	case BlockInt64:
+		return decodeInt64Block(block)
+	case BlockBool:
+		return decodeBoolBlock(block)
+	case BlockString:
+		return decodeStringBlock(block)
+	default:
+	}
+
+	// TODO: add support for other block types
+	return nil, fmt.Errorf("unknown block type: %d", blockType)
+}
+
+// Deduplicate returns a new Values slice with any values
+// that have the same  timestamp removed. The Value that appears
+// last in the slice is the one that is kept. The returned slice is in ascending order
+func (v Values) Deduplicate() Values {
+	m := make(map[int64]Value)
+	for _, val := range v {
+		m[val.UnixNano()] = val
+	}
+
+	a := make([]Value, 0, len(m))
+	for _, val := range m {
+		a = append(a, val)
+	}
+	sort.Sort(Values(a))
+
+	return a
+}
+
+// Sort methods
+func (a Values) Len() int           { return len(a) }
+func (a Values) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a Values) Less(i, j int) bool { return a[i].Time().UnixNano() < a[j].Time().UnixNano() }
+
+type FloatValue struct {
+	time  time.Time
+	value float64
+}
+
+func (f *FloatValue) Time() time.Time {
+	return f.time
+}
+
+func (f *FloatValue) UnixNano() int64 {
+	return f.time.UnixNano()
+}
+
+func (f *FloatValue) Value() interface{} {
+	return f.value
+}
+
+func (f *FloatValue) Size() int {
+	return 16
```

Comment:
would a const actually add any extra detail? I don't think so given that it's a single line method. I could change it to `sizeof(int64) + sizeof(float64)`, but that feels lame to me. Unnecessary math...